docs/guide/generics/README.md

```diff
@@ -0,0 +1,146 @@
+
+Usually, you use generic nodes like regular ones and don’t think of logic behind
+type deduction, and resolution process. Although to handle some edge-cases you
+need to understand principles described below.
+
+## Resolution process overview
+
+Whenever you transpile a XOD program which contains generic nodes, the system
+_resolves_ them to concrete _specializations_, so that a particular C++
+implementation can be found in the final turn.
+
+The system tries to replace in-place each generic node with another node called
+a _specialization node_ which is chosen based on _deduced_ input types.
+
+A generic node can be a composition of other generic nodes, or it can be
+_abstract_. Abstract nodes can’t translate to C++ because they contain no
+implementation and serve as placeholders for the transpiler (hence called
+abstract). If XOD can’t find a required specialization for an abstract node, it
+shows an error.
+
+When dealing with a generic composite node, the system searches for a matching
+specialization, and if not found, it expands the composition and reruns
+resolution. The process repeats recursively until all specializations are found
+or only unresolved abstract nodes left, which is an error.
+
+Let’s elaborate on each of concepts mentioned.
+
+## Specializations
+
+A specialization is a regular patch having a distinctive name like:
+
+* `if-else(number)`
+* `if-else(string)`
+
+The name is made up of the abstraction patch name (`if-else` in this case) and
+type specialization for `t1` enclosed in parentheses.
+
+If a generic node has two or more generic types, the types in parens should be
+comma-separated, e.g., `foo(string,number,pulse)`.
+
+In most cases, you should not bother on the specialization patches and nodes as
+you’re interfacing with the base abstraction (`if-else`) and that’s enough. The
+specializations don’t even appear in the quick-search suggester by default.
+
+However, sometimes you will want to be specific and use a specialization
+directly, like a regular node. That’s called _manual resolution_. To access a
+particular specialization, use one of the following methods:
+
+1. Find the specialization in the Project Browser and drag-n-drop it on your
+   patch as you would do with any regular node.
+2. Double-click your patch, type a part of the node name and add a left
+   parenthesis to see the specializations (e.g., type `if-el(`).
+3. Select a generic node to resolve manually, then in the inspector choose the
+   desired specialization in the drop-down below the node name.
+
+![Screencast](./screencast.gif)
+
+## Type deduction rules
+
+To be successfully resolved later a generic node should match a specific type to
+each of `t1`, `t2`, and `t3`. It is called _type deduction_.
+
+The deduction is done based on actual types of links and values bound to the
+generic pins. XOD will show an error if types can’t be deduced unambiguously,
+for example when one pin of type `t1` pretend to be number and another `t1` pin
```

Comment:
Maybe add a hint (or a separate subsection), that if you want to use two different types in a generic node with casting one of the inputs, place cast nodes manually (`formatNumber`, `equals` and etc)?